File: lib/scraper/src/api/steam.ts
import { z } from 'zod';
import { escapeMogDelimiters, mogImage, mogLink } from '@deadlog/utils';

const STEAM_NEWS_API = 'https://api.steampowered.com/ISteamNews/GetNewsForApp/v2/';
const DEADLOCK_APP_ID = '1422450';

const SECTION_MARKER = '__SECTION__';
const SUBPATCH_MARKER = '__SUBPATCH__';
const SECTION_NAMES = new Set(['Items', 'Heroes', 'General']);
const STEAM_IMAGE_BASE = 'https://clan.fastly.steamstatic.com/images';
const STEAM_IMAGE_RE =
	/\[img(?:\s+src=(?:"([^"]+)"|'([^']+)'|([^\]\s]+)))?\]([\s\S]*?)\[\/img\]/gi;

const steamNewsItemSchema = z.object({
	gid: z.string(),
	title: z.string(),
	url: z.string(),
	author: z.string(),
	contents: z.string(),
	feedlabel: z.string(),
	date: z.number(),
	feedname: z.string(),
	feed_type: z.number(),
	appid: z.number(),
	tags: z.array(z.string()).optional()
});

const steamNewsResponseSchema = z.object({
	appnews: z.object({
		appid: z.number(),
		newsitems: z.array(steamNewsItemSchema)
	})
});

export type SteamNewsItem = z.infer<typeof steamNewsItemSchema>;

export interface SteamAnnouncement {
	gid: string;
	title: string;
	date: string; // ISO date string
	author: string;
	content: string; // BBCode content
}

export interface RenderedSteamAnnouncement {
	mog: string;
	text: string;
}

/** MM-DD-YYYY out of "Gameplay Update - 03-06-2026" or "03-06-2026 Update". */
export function extractDateFromTitle(title: string): string | null {
	const match = title.match(/(\d{2}-\d{2}-\d{4})/);
	return match ? match[1] : null;
}

function bbcodeToText(bbcode: string): string {
	let text = bbcode;

	// Plain-text parsing omits presentation-only media.
	text = text
		.replace(/\[img(?:\s+[^\]]*)?\][\s\S]*?\[\/img\]/gi, '\n')
		.replace(/\{STEAM_CLAN_(?:LOC_)?IMAGE\}\/\S+/g, '');

	text = text
		.replace(/\[url=[^\]]+\]([\s\S]*?)\[\/url\]/gi, '$1')
		.replace(/\[url\]([\s\S]*?)\[\/url\]/gi, '$1')
		.replace(/\[video(?:\s+[^\]]*)?\][\s\S]*?\[\/video\]/gi, '\n')
		.replace(/\[\*\]\s*/g, '\n- ');

	// [/p][p] is a single newline; a standalone tag is a newline or nothing.
	text = text
		.replace(/\[\/p\]\[p\]/g, '\n')
		.replace(/\[\/?p\]/g, (m) => (m === '[p]' ? '' : '\n'));
	text = text.replace(/\[\/?(?:h[1-6]|list)\]/gi, '\n');

	// Section header shape: [u][b]\[ General ][/b][/u]
	text = text.replace(
		/\[u\]\[b\]\\?\[\s*([^\]]+?)\s*\]?\[\/b\]\[\/u\]/g,
		`${SECTION_MARKER}$1`
	);
	text = text.replace(/\[b\]\\?\[\s*([^\]]+?)\s*\]?\[\/b\]/g, (_, name) => {
		const trimmed = name.trim();
		if (SECTION_NAMES.has(trimmed)) {
			return `${SECTION_MARKER}${trimmed}`;
		}
		// Sub-patch headers like "03-07-2026 Patch:"
		if (/^\d{2}-\d{2}-\d{4}\s+Patch/i.test(trimmed)) {
			return `${SUBPATCH_MARKER}${trimmed}`;
		}
		return trimmed;
	});

	text = text.replace(/\[\/?\w+(?:[=\s][^\]]*)?\]/g, '');

	text = text.replace(/\\(?=\[)/g, '');

	text = text.replace(/\n{3,}/g, '\n\n');

	return text.trim();
}

/** Plain-text lines, shaped to feed parseAndGroupContent directly. */
export function parseSteamContent(bbcode: string): string {
	const processed = bbcodeToText(bbcode);
	const lines = processed.split('\n');
	const output: string[] = [];

	for (const line of lines) {
		const trimmed = line.trim();
		if (!trimmed) continue;

		if (trimmed.startsWith(SECTION_MARKER) || trimmed.startsWith(SUBPATCH_MARKER)) {
			continue;
		}

		output.push(trimmed);
	}

	return output.join('\n');
}

export function isSteamPatchContent(bbcode: string): boolean {
	const withoutBbcodeLists = bbcode.replace(/\[list\][\s\S]*?\[\/list\]/gi, '');
	return parseSteamContent(withoutBbcodeLists)
		.split('\n')
		.some((line) => /^(?:[-•]\s*\S|\*\s+\S)/.test(line.trim()));
}

function safeHttpUrl(value: string): string | null {
	try {
		const url = new URL(value.trim().replace(/^"|"$/g, ''));
		return ['http:', 'https:'].includes(url.protocol) && !/[[\]()]/.test(url.href)
			? url.href
			: null;
	} catch {
		return null;
	}
}

function steamImageUrl(value: string): string | null {
	return safeHttpUrl(
		value.trim().replace(/^\{STEAM_CLAN_(?:LOC_)?IMAGE\}/i, STEAM_IMAGE_BASE)
	);
}

function plainBbcode(value: string): string {
	return value
		.replace(/\[\/?\w+(?:[=\s][^\]]*)?\]/g, '')
		.replace(/\\(?=\[)/g, '')
		.replace(/\s+/g, ' ')
		.trim();
}

function safeMogLabel(value: string): string {
	return plainBbcode(value)
		.replace(/[[\]()]/g, '')
		.trim();
}

/** Convert prose-shaped Steam BBCode directly to readable Mog blocks. */
export function renderSteamAnnouncement(
	title: string,
	bbcode: string
): RenderedSteamAnnouncement {
	const tokens: string[] = [];
	const protect = (value: string) => {
		const index = tokens.push(value) - 1;
		return `\uE000${index}\uE001`;
	};
	const restore = (value: string) =>
		value.replace(/\uE000(\d+)\uE001/g, (_, index) => tokens[Number(index)] ?? '');
	const hasExplicitHeading = /\[h[1-6]\]/i.test(bbcode);

	let content = bbcode.replace(/\r\n?/g, '\n').replace(/\u00a0/g, ' ');

	content = content.replace(
		/\[url=([^\]]+)\]([\s\S]*?)\[\/url\]/gi,
		(_match, target: string, label: string) => {
			if (/\[img(?:\s|\])/i.test(label)) return label;
			const href = safeHttpUrl(target);
			const text = safeMogLabel(label);
			return href && text ? protect(mogLink(href, text)) : text;
		}
	);

	content = content.replace(
		STEAM_IMAGE_RE,
		(_match, doubleQuoted: string, singleQuoted: string, bare: string, body: string) => {
			const src = steamImageUrl(doubleQuoted || singleQuoted || bare || body);
			if (!src) return '\n';
			return `\n${protect(mogImage(src, `${safeMogLabel(title)} announcement art`))}\n`;
		}
	);

	content = content.replace(
		/\[video\s+[^\]]*\bmp4=(?:"([^"]+)"|'([^']+)'|([^\]\s]+))[^\]]*\][\s\S]*?\[\/video\]/gi,
		(_match, doubleQuoted: string, singleQuoted: string, bare: string) => {
			const href = safeHttpUrl(doubleQuoted || singleQuoted || bare);
			return href ? `\n${protect(mogLink(href, 'Video'))}\n` : '\n';
		}
	);

	content = content
		.replace(/\[h[1-6]\]([\s\S]*?)\[\/h[1-6]\]/gi, '\n\uE100$1\n')
		.replace(/\[\/?p\]/gi, '\n')
		.replace(/\[\/?list\]/gi, '\n')
		.replace(/\[\*\]\s*/gi, '\n\uE101')
		.replace(/\[br\s*\/?\]/gi, '\n');

	const formatInline = (value: string): string => {
		let text = value
			.replace(/\[i\]([\s\S]*?)\[\/i\]/gi, (_match, label: string) =>
				protect(`__${escapeMogDelimiters(plainBbcode(label))}__`)
			)
			.replace(/\[b\]([\s\S]*?)\[\/b\]/gi, (_match, label: string) =>
				protect(`**${escapeMogDelimiters(plainBbcode(label))}**`)
			)
			.replace(/\[\/?u\]/gi, '')
			.replace(/\[\/?\w+(?:[=\s][^\]]*)?\]/g, '')
			.replace(/\\(?=\[)/g, '')
			.replace(/\s+/g, ' ')
			.trim();
		text = escapeMogDelimiters(text);
		if (/^[#=.`>|+*~:$-]/.test(text)) text = `\\${text}`;
		return restore(text);
	};

	const blocks: { value: string; list: boolean }[] = [];
	const append = (value: string, list = false) => {
		if (!value) return;
		const previous = blocks.at(-1);
		if (list && previous?.list) previous.value += `\n${value}`;
		else blocks.push({ value, list });
	};

	for (const rawLine of content.split('\n')) {
		const line = rawLine.trim();
		if (!line || line === '-') continue;

		if (line.startsWith('\uE100')) {
			append(`## ${formatInline(line.slice(1))}`);
			continue;
		}

		if (line.startsWith('\uE101')) {
			append(`- ${formatInline(line.slice(1))}`, true);
			continue;
		}

		const boldHeading = line.match(/^\[b\](.*?)\[\/b\]$/i);
		if (boldHeading) {
			append(`${hasExplicitHeading ? '###' : '#'} ${formatInline(boldHeading[1])}`);
			continue;
		}

		append(formatInline(line));
	}

	const text = parseSteamContent(bbcode)
		.split('\n')
		.map((line) => line.replace(/^(?:[-*•]+\s*)/, '').trim())
		.filter(Boolean)
		.join(' ');

	return { mog: blocks.map((block) => block.value).join('\n\n'), text };
}

export async function fetchSteamAnnouncements(options: {
	count?: number;
	timeout?: number;
}): Promise<SteamAnnouncement[]> {
	const { count = 1000, timeout = 30000 } = options;

	const url = new URL(STEAM_NEWS_API);
	url.searchParams.set('appid', DEADLOCK_APP_ID);
	url.searchParams.set('count', String(count));
	url.searchParams.set('maxlength', '0'); // 0 = full content
	url.searchParams.set('feeds', 'steam_community_announcements');

	const response = await fetch(url.toString(), {
		signal: AbortSignal.timeout(timeout)
	});

	if (!response.ok) {
		throw new Error(`Steam API HTTP ${response.status}: ${response.statusText}`);
	}

	const data = steamNewsResponseSchema.parse(await response.json());

	return data.appnews.newsitems.map((item) => ({
		gid: item.gid,
		title: item.title,
		date: new Date(item.date * 1000).toISOString(),
		author: item.author,
		content: item.contents
	}));
}

export function extractSteamUrlFromUnfurl(html: string): string | null {
	const match = html.match(
		/href="(https:\/\/store\.steampowered\.com\/news\/app\/\d+\/view\/\d+)"/
	);
	return match ? match[1] : null;
}

/** True when the post is a link preview with no patch notes of its own. */
export function isSteamUnfurl(html: string): boolean {
	if (!extractSteamUrlFromUnfurl(html)) return false;
	if (!html.includes('bbCodeBlock--unfurl')) return false;

	const stripped = html
		.replace(
			/<div class="bbCodeBlock bbCodeBlock--unfurl[\s\S]*?<\/div>\s*<\/div>\s*<\/div>/g,
			''
		)
		.replace(/<[^>]+>/g, '')
		.trim();

	return stripped.length < 50;
}

File: crates/scraper/src/api/steam.rs
use std::sync::LazyLock;
use std::time::Duration;

use anyhow::{Context, Result, bail};
use chrono::{TimeZone, Utc};
use deadlog_model::{JS_SPACE, escape_mog_delimiters, js_trim, mog_image, mog_link};
use regex::{Captures, Regex};
use serde_json::Value;

use crate::http::Http;
use deadlog_model::utf16_len;

const STEAM_NEWS_API: &str = "https://api.steampowered.com/ISteamNews/GetNewsForApp/v2/";
const DEADLOCK_APP_ID: &str = "1422450";

const SECTION_MARKER: &str = "__SECTION__";
const SUBPATCH_MARKER: &str = "__SUBPATCH__";
const SECTION_NAMES: [&str; 3] = ["Items", "Heroes", "General"];
const STEAM_IMAGE_BASE: &str = "https://clan.fastly.steamstatic.com/images";

#[derive(Debug, Clone, PartialEq)]
pub struct SteamAnnouncement {
    pub gid: String,
    pub title: String,
    /// ISO timestamp.
    pub date: String,
    pub author: String,
    /// BBCode.
    pub content: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct RenderedSteamAnnouncement {
    pub mog: String,
    pub text: String,
}

fn re(pattern: &str) -> Regex {
    Regex::new(&pattern.replace(r"\s", JS_SPACE).replace(r"\S", &format!("[^{}]", &JS_SPACE[1..JS_SPACE.len() - 1])))
        .unwrap()
}

macro_rules! regex {
    ($name:ident, $pattern:expr) => {
        static $name: LazyLock<Regex> = LazyLock::new(|| re($pattern));
    };
}

regex!(TITLE_DATE, r"([0-9]{2}-[0-9]{2}-[0-9]{4})");
regex!(IMG_ANY, r"(?i)\[img(?:\s+[^\]]*)?\][\s\S]*?\[/img\]");
regex!(CLAN_IMAGE_TOKEN, r"\{STEAM_CLAN_(?:LOC_)?IMAGE\}/\S+");
regex!(URL_WITH_TARGET, r"(?i)\[url=[^\]]+\]([\s\S]*?)\[/url\]");
regex!(URL_BARE, r"(?i)\[url\]([\s\S]*?)\[/url\]");
regex!(VIDEO_ANY, r"(?i)\[video(?:\s+[^\]]*)?\][\s\S]*?\[/video\]");
regex!(LIST_ITEM, r"\[\*\]\s*");
regex!(PARAGRAPH, r"\[/?p\]");
regex!(HEADING_OR_LIST, r"(?i)\[/?(?:h[1-6]|list)\]");
regex!(UNDERLINED_SECTION, r"\[u\]\[b\]\\?\[\s*([^\]]+?)\s*\]?\[/b\]\[/u\]");
regex!(BOLD_SECTION, r"\[b\]\\?\[\s*([^\]]+?)\s*\]?\[/b\]");
regex!(SUBPATCH, r"(?i)^[0-9]{2}-[0-9]{2}-[0-9]{4}\s+Patch");
regex!(ANY_TAG, r"\[/?[A-Za-z0-9_]+(?:[=\s][^\]]*)?\]");
regex!(ESCAPED_BRACKET, r"\\\[");
regex!(BLANK_RUNS, r"\n{3,}");
regex!(BBCODE_LIST, r"(?i)\[list\][\s\S]*?\[/list\]");
regex!(PATCH_BULLET, r"^(?:[-•]\s*\S|\*\s+\S)");
regex!(SPACES, r"\s+");
regex!(QUOTES, r#"^"|"$"#);
regex!(CLAN_IMAGE_PREFIX, r"(?i)^\{STEAM_CLAN_(?:LOC_)?IMAGE\}");
regex!(LINK_DELIMITERS, r"[\[\]()]");
regex!(EXPLICIT_HEADING, r"(?i)\[h[1-6]\]");
regex!(LINE_BREAKS, r"\r\n?");
regex!(URL_TAG, r"(?i)\[url=([^\]]+)\]([\s\S]*?)\[/url\]");
regex!(IMG_IN_LABEL, r"(?i)\[img(?:\s|\])");
regex!(STEAM_IMAGE, r#"(?i)\[img(?:\s+src=(?:"([^"]+)"|'([^']+)'|([^\]\s]+)))?\]([\s\S]*?)\[/img\]"#);
regex!(VIDEO_MP4, r#"(?i)\[video\s+[^\]]*(?-u:\b)mp4=(?:"([^"]+)"|'([^']+)'|([^\]\s]+))[^\]]*\][\s\S]*?\[/video\]"#);
regex!(HEADING_TAG, r"(?i)\[h[1-6]\]([\s\S]*?)\[/h[1-6]\]");
regex!(PARAGRAPH_ANY_CASE, r"(?i)\[/?p\]");
regex!(LIST_TAG, r"(?i)\[/?list\]");
regex!(LIST_ITEM_ANY_CASE, r"(?i)\[\*\]\s*");
regex!(BREAK_TAG, r"(?i)\[br\s*/?\]");
regex!(ITALIC, r"(?i)\[i\]([\s\S]*?)\[/i\]");
regex!(BOLD, r"(?i)\[b\]([\s\S]*?)\[/b\]");
regex!(UNDERLINE, r"(?i)\[/?u\]");
regex!(BLOCK_OPENER, r"^[#=.`>|+*~:$-]");
regex!(BOLD_LINE, r"(?i)^\[b\](.*?)\[/b\]$");
regex!(TOKEN, "\u{E000}([0-9]+)\u{E001}");
regex!(LEADING_MARKERS, r"^(?:[-*•]+\s*)");
regex!(STEAM_VIEW_URL, r#"href="(https://store\.steampowered\.com/news/app/[0-9]+/view/[0-9]+)""#);
regex!(UNFURL_BLOCK, r#"<div class="bbCodeBlock bbCodeBlock--unfurl[\s\S]*?</div>\s*</div>\s*</div>"#);
regex!(HTML_TAG, r"<[^>]+>");

/// `MM-DD-YYYY` out of "Gameplay Update - 03-06-2026" or "03-06-2026 Update".
pub fn extract_date_from_title(title: &str) -> Option<String> {
    TITLE_DATE.captures(title).map(|captures| captures[1].to_string())
}

fn bbcode_to_text(bbcode: &str) -> String {
    // Plain-text parsing omits presentation-only media.
    let text = IMG_ANY.replace_all(bbcode, "\n");
    let text = CLAN_IMAGE_TOKEN.replace_all(&text, "");
    let text = URL_WITH_TARGET.replace_all(&text, "$1");
    let text = URL_BARE.replace_all(&text, "$1");
    let text = VIDEO_ANY.replace_all(&text, "\n");
    let text = LIST_ITEM.replace_all(&text, "\n- ");

    // [/p][p] is a single newline; a standalone tag is a newline or nothing.
    let text = text.replace("[/p][p]", "\n");
    let text = PARAGRAPH.replace_all(&text, |captures: &Captures| if &captures[0] == "[p]" { "" } else { "\n" });
    let text = HEADING_OR_LIST.replace_all(&text, "\n");

    // Section header shape: [u][b]\[ General ][/b][/u]
    let text = UNDERLINED_SECTION.replace_all(&text, format!("{SECTION_MARKER}${{1}}").as_str());
    let text = BOLD_SECTION.replace_all(&text, |captures: &Captures| {
        let trimmed = js_trim(&captures[1]);
        if SECTION_NAMES.contains(&trimmed) {
            return format!("{SECTION_MARKER}{trimmed}");
        }
        // Sub-patch headers like "03-07-2026 Patch:"
        if SUBPATCH.is_match(trimmed) {
            return format!("{SUBPATCH_MARKER}{trimmed}");
        }
        trimmed.to_string()
    });

    let text = ANY_TAG.replace_all(&text, "");
    let text = ESCAPED_BRACKET.replace_all(&text, "[");
    let text = BLANK_RUNS.replace_all(&text, "\n\n");
    js_trim(&text).to_string()
}

/// Plain-text lines, shaped to feed `parse_and_group_content` directly.
pub fn parse_steam_content(bbcode: &str) -> String {
    bbcode_to_text(bbcode)
        .split('\n')
        .map(js_trim)
        .filter(|line| !line.is_empty() && !line.starts_with(SECTION_MARKER) && !line.starts_with(SUBPATCH_MARKER))
        .collect::<Vec<_>>()
        .join("\n")
}

pub fn is_steam_patch_content(bbcode: &str) -> bool {
    let without_lists = BBCODE_LIST.replace_all(bbcode, "");
    parse_steam_content(&without_lists).split('\n').any(|line| PATCH_BULLET.is_match(js_trim(line)))
}

fn safe_http_url(value: &str) -> Option<String> {
    let url = url::Url::parse(&QUOTES.replace_all(js_trim(value), "")).ok()?;
    let href = url.as_str();
    (matches!(url.scheme(), "http" | "https") && !LINK_DELIMITERS.is_match(href)).then(|| href.to_string())
}

fn steam_image_url(value: &str) -> Option<String> {
    safe_http_url(&CLAN_IMAGE_PREFIX.replace(js_trim(value), STEAM_IMAGE_BASE))
}

fn plain_bbcode(value: &str) -> String {
    let text = ANY_TAG.replace_all(value, "");
    let text = ESCAPED_BRACKET.replace_all(&text, "[");
    js_trim(&SPACES.replace_all(&text, " ")).to_string()
}

fn safe_mog_label(value: &str) -> String {
    js_trim(&LINK_DELIMITERS.replace_all(&plain_bbcode(value), "")).to_string()
}

fn first_group<'a>(captures: &'a Captures, groups: &[usize]) -> &'a str {
    groups
        .iter()
        .filter_map(|index| captures.get(*index))
        .map(|m| m.as_str())
        .find(|value| !value.is_empty())
        .unwrap_or("")
}

#[derive(Default)]
struct Tokens(std::cell::RefCell<Vec<String>>);

impl Tokens {
    fn protect(&self, value: String) -> String {
        let mut tokens = self.0.borrow_mut();
        tokens.push(value);
        format!("\u{E000}{}\u{E001}", tokens.len() - 1)
    }

    fn restore(&self, value: &str) -> String {
        let tokens = self.0.borrow();
        TOKEN
            .replace_all(value, |captures: &Captures| {
                captures[1].parse::<usize>().ok().and_then(|index| tokens.get(index)).cloned().unwrap_or_default()
            })
            .into_owned()
    }
}

/// Convert prose-shaped Steam BBCode directly to readable Mog blocks.
pub fn render_steam_announcement(title: &str, bbcode: &str) -> RenderedSteamAnnouncement {
    let tokens = Tokens::default();
    let has_explicit_heading = EXPLICIT_HEADING.is_match(bbcode);

    let content = LINE_BREAKS.replace_all(bbcode, "\n").replace('\u{a0}', " ");
    let content = URL_TAG.replace_all(&content, |captures: &Captures| {
        let (target, label) = (&captures[1], &captures[2]);
        if IMG_IN_LABEL.is_match(label) {
            return label.to_string();
        }
        let text = safe_mog_label(label);
        match safe_http_url(target) {
            Some(href) if !text.is_empty() => tokens.protect(mog_link(&href, &text)),
            _ => text,
        }
    });
    let content = STEAM_IMAGE.replace_all(&content, |captures: &Captures| {
        match steam_image_url(first_group(captures, &[1, 2, 3, 4])) {
            None => "\n".to_string(),
            Some(src) => format!(
                "\n{}\n",
                tokens.protect(mog_image(&src, &format!("{} announcement art", safe_mog_label(title))))
            ),
        }
    });
    let content =
        VIDEO_MP4.replace_all(&content, |captures: &Captures| match safe_http_url(first_group(captures, &[1, 2, 3])) {
            Some(href) => format!("\n{}\n", tokens.protect(mog_link(&href, "Video"))),
            None => "\n".to_string(),
        });
    let content = HEADING_TAG.replace_all(&content, "\n\u{E100}${1}\n");
    let content = PARAGRAPH_ANY_CASE.replace_all(&content, "\n");
    let content = LIST_TAG.replace_all(&content, "\n");
    let content = LIST_ITEM_ANY_CASE.replace_all(&content, "\n\u{E101}");
    let content = BREAK_TAG.replace_all(&content, "\n");

    let format_inline = |value: &str| -> String {
        let text = ITALIC.replace_all(value, |captures: &Captures| {
            tokens.protect(format!("__{}__", escape_mog_delimiters(&plain_bbcode(&captures[1]))))
        });
        let text = BOLD.replace_all(&text, |captures: &Captures| {
            tokens.protect(format!("**{}**", escape_mog_delimiters(&plain_bbcode(&captures[1]))))
        });
        let text = UNDERLINE.replace_all(&text, "");
        let text = ANY_TAG.replace_all(&text, "");
        let text = ESCAPED_BRACKET.replace_all(&text, "[");
        let text = SPACES.replace_all(&text, " ");
        let mut text = escape_mog_delimiters(js_trim(&text));
        if BLOCK_OPENER.is_match(&text) {
            text = format!("\\{text}");
        }
        tokens.restore(&text)
    };

    let mut blocks: Vec<(String, bool)> = Vec::new();
    let mut append = |value: String, list: bool| {
        if value.is_empty() {
            return;
        }
        match blocks.last_mut() {
            Some((previous, true)) if list => {
                previous.push('\n');
                previous.push_str(&value);
            }
            _ => blocks.push((value, list)),
        }
    };

    for raw_line in content.split('\n') {
        let line = js_trim(raw_line);
        if line.is_empty() || line == "-" {
            continue;
        }
        if let Some(heading) = line.strip_prefix('\u{E100}') {
            append(format!("## {}", format_inline(heading)), false);
        } else if let Some(item) = line.strip_prefix('\u{E101}') {
            append(format!("- {}", format_inline(item)), true);
        } else if let Some(captures) = BOLD_LINE.captures(line) {
            let marker = if has_explicit_heading { "###" } else { "#" };
            append(format!("{marker} {}", format_inline(&captures[1])), false);
        } else {
            append(format_inline(line), false);
        }
    }

    let text = parse_steam_content(bbcode)
        .split('\n')
        .map(|line| js_trim(&LEADING_MARKERS.replace(line, "")).to_string())
        .filter(|line| !line.is_empty())
        .collect::<Vec<_>>()
        .join(" ");

    RenderedSteamAnnouncement { mog: blocks.into_iter().map(|(value, _)| value).collect::<Vec<_>>().join("\n\n"), text }
}

fn string_field<'a>(item: &'a Value, key: &str) -> Result<&'a str> {
    item.get(key).and_then(Value::as_str).with_context(|| format!("newsitems: {key} must be a string"))
}

/// Every Deadlock community announcement Steam still serves, full text, newest first.
pub fn fetch_steam_announcements(http: &dyn Http, count: usize, timeout: Duration) -> Result<Vec<SteamAnnouncement>> {
    let url = format!(
        "{STEAM_NEWS_API}?appid={DEADLOCK_APP_ID}&count={count}&maxlength=0&feeds=steam_community_announcements"
    );
    let response = http.get(&url, &[], Some(timeout))?;
    if !response.ok() {
        bail!("Steam API HTTP {}: {}", response.status, response.status_text);
    }
    let data: Value = serde_json::from_str(&response.body)?;
    let appnews = data.get("appnews").context("appnews missing")?;
    appnews.get("appid").and_then(Value::as_f64).context("appnews.appid must be a number")?;
    let items = appnews.get("newsitems").and_then(Value::as_array).context("appnews.newsitems must be an array")?;
    items
        .iter()
        .map(|item| {
            for key in ["url", "feedlabel", "feedname"] {
                string_field(item, key)?;
            }
            for key in ["feed_type", "appid"] {
                item.get(key).and_then(Value::as_f64).with_context(|| format!("newsitems: {key} must be a number"))?;
            }
            if let Some(tags) = item.get("tags")
                && !tags.as_array().is_some_and(|tags| tags.iter().all(Value::is_string))
            {
                bail!("newsitems: tags must be strings");
            }
            let seconds = item.get("date").and_then(Value::as_f64).context("newsitems: date must be a number")?;
            let date = Utc
                .timestamp_millis_opt((seconds * 1000.0) as i64)
                .single()
                .context("newsitems: date out of range")?
                .format("%Y-%m-%dT%H:%M:%S%.3fZ")
                .to_string();
            Ok(SteamAnnouncement {
                gid: string_field(item, "gid")?.to_string(),
                title: string_field(item, "title")?.to_string(),
                date,
                author: string_field(item, "author")?.to_string(),
                content: string_field(item, "contents")?.to_string(),
            })
        })
        .collect()
}

/// True when the post is a link preview with no patch notes of its own.
pub fn is_steam_unfurl(html: &str) -> bool {
    if !STEAM_VIEW_URL.is_match(html) || !html.contains("bbCodeBlock--unfurl") {
        return false;
    }
    let stripped = UNFURL_BLOCK.replace_all(html, "");
    let stripped = HTML_TAG.replace_all(&stripped, "");
    utf16_len(js_trim(&stripped)) < 50
}
